### PR TITLE
[codex] recover watcher and hosted review failures

### LINKED ISSUE #8963: Recover filesystem watcher overflows and hosted-review lookup failures
## Problem

Two background recovery paths can leave Orca noisy or stale:

1. On macOS, `@parcel/watcher` can report that FSEvents dropped events and a full rescan is required. Orca currently forwards this recoverable condition as a terminal watcher error, causing the desktop filesystem watcher to unsubscribe and silently stop delivering later external file changes.
2. Hosted-review lookup probes GitHub repository candidates in upstream-then-origin order. A stale, deleted, private, or otherwise unavailable upstream candidate can abort lookup before a valid origin is probed. Visible-card polling and renderer HMR repeatedly throw the resulting `repo_unavailable` failure through Electron IPC.

Other forge providers can similarly collapse transient CLI or HTTP failures into a definitive “no review” result, causing cached review state to disappear temporarily.

## Acceptance criteria

- The known macOS dropped-FSEvents condition produces a conservative full refresh without destroying a healthy subscription.
- Later file events continue after overflow on desktop, runtime, and relay watcher paths.
- Genuine watcher errors replace or terminate the untrusted subscription through bounded recovery rather than remaining silently live.
- A stale or inaccessible upstream GitHub candidate does not prevent probing a valid origin candidate.
- If all repository candidates fail, callers receive a safe typed upstream failure rather than a false “no review” result.
- Desktop IPC and paired-runtime RPC use the same provider-generic result contract.
- GitHub, GitLab, Bitbucket, Gitea, and Azure DevOps distinguish genuine review misses from transient operational failures.
- The renderer preserves the last cached review on transient failure and retries later.
- Diagnostics are structured, contain only safe opaque context, survive HMR for deduplication, and reset after recovery or an error-signature change.
- Older compatible runtimes returning `HostedReviewInfo | null` continue to work.
- Focused regression coverage exercises candidate fallback, provider failure modes, watcher continuity/replacement, transport contracts, cache preservation, and diagnostic deduplication.

### PR DESCRIPTION
**Pull request by morluto**
Closes #8963

## Summary

Make background watcher and hosted-review failures recover without silently losing state or producing repeated Electron IPC exceptions.

- Treat macOS's known dropped-FSEvents callback as a live overflow: request a conservative refresh while retaining the subscription and later event delivery.
- Make runtime and relay watcher owners replace genuinely untrusted subscriptions through generation-safe bounded recovery. Duplicate callbacks and stale recovery failures cannot retire a newer healthy generation.
- Continue GitHub PR candidate lookup after a stale, private, deleted, or unavailable upstream candidate so a valid origin can still succeed.
- Return provider-generic found/not-found/typed-upstream-error outcomes across desktop IPC and paired-runtime RPC, while normalizing legacy compatible runtime responses.
- Preserve cached review state and retry after transient failures; persist safe diagnostic deduplication through renderer HMR.
- Distinguish genuine misses from transient operational failures across GitHub, GitLab, Bitbucket, Gitea, and Azure DevOps using one shared request failure policy and bounded safe `Error.cause` classification.

## Screenshots

No visual change.

## Testing

- [x] `pnpm lint`
- [x] `pnpm typecheck`
- [ ] `pnpm test` — all 18 changed suites passed, totaling 1,050 tests; the full unrelated suite was not run.
- [ ] `pnpm build` — no production artifact format changed; full build was not run.
- [x] Added or updated high-quality tests covering FSEvents overflow continuity, terminal watcher replacement and races, GitHub candidate fallback, provider failure modes, IPC/RPC contracts, legacy responses, cache preservation, and HMR diagnostic deduplication.

## AI Review Report

The branch received repeated independent Standards and Spec reviews plus structured Codex autoreview against `origin/main`. Reviews explicitly checked macOS/Linux/Windows watcher behavior, SSH and paired-runtime parity, GitHub/GitLab/Bitbucket/Gitea/Azure provider semantics, legacy runtime compatibility, safe diagnostics, generation ownership, and cross-platform path/runtime behavior.

Accepted findings fixed during review included provider discovery outside the typed boundary, duplicated transport result projection, duplicated cache fallback, non-GitHub transient failures becoming false misses, terminal watcher parity gaps, HMR-local diagnostic deduplication, legacy runtime result compatibility, native `fetch failed` cause-chain classification, repeated provider failure-policy primitives, and duplicate/stale watcher recovery races. Final Standards, Spec, and structured autoreview passes are clean with no actionable findings.

## Security Audit

Reviewed provider error handling, IPC/RPC serialization, logs, external CLI/HTTP failure boundaries, filesystem watcher ownership, and retry behavior. Diagnostics contain only bounded provider/error tags and opaque repository identifiers; branch names, paths, remotes, raw errors, credentials, and nested causes are not logged. Error-cause traversal is bounded and projects only stable network codes. No new shell interpolation, command construction, authentication flow, dependency, or persisted secret handling was introduced.

## Notes

- The original `repo_unavailable` loop was caused by upstream-first candidate probing aborting before a valid origin lookup; visible-card polling and HMR amplified the same thrown IPC error.
- Compatible older runtime servers returning `HostedReviewInfo | null` remain supported at the renderer boundary.
- Verification and commit hooks used the repository-supported Node 24.18.0 runtime; the local Homebrew Node 25 installation is unsupported and currently missing `libsimdjson.29.dylib`.

## ELI5

Background file watchers and hosted review could fail silently or spam Electron errors. They recover safely now—refreshing or replacing broken watches without losing ongoing state.

### CODE REVIEW

File: src/shared/watcher-generation-replacement.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  const release = onClaim()
+  const promise = (async () => {
+    if (release) {
+      await release()
+    }
+    if (owner.closed || owner.generation !== generation) {
+      return
+    }
+    await install(generation)
+  })()
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/shared/watcher-generation-replacement.ts:32]: Addressed in eb6818303. The claim hook now runs inside the tracked async replacement body, so synchronous failures reject replacement.promise; a focused regression test verifies generation ownership and skipped install.

File: src/main/source-control/hosted-review.ts [hunk 2]
```diff
@@ -1,9 +1,93 @@
...
+  'UND_ERR_SOCKET'
+])
+
+function readErrorProperty(error: object, property: 'cause' | 'code' | 'message'): unknown {
+  try {
+    return Reflect.get(error, property)
+  } catch {
+    return undefined
+  }
+}
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/main/source-control/hosted-review.ts:98]: Addressed in eb6818303. Provider error traversal now records error names across the bounded cause chain and classifies standard AbortError failures as network errors, with top-level and nested-cause tests.

File: src/relay/relay-watcher-recovery.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
+import type { RelayDispatcher } from './dispatcher'
+import { emitRelayWatcherOverflow } from './relay-watcher-event-emitter'
+import { emitRelayWatcherTerminalFailure } from './relay-watcher-terminal-notifier'
+import type { RelayWatcherTeardownState } from './relay-watcher-teardown-tracker'
+import { PromiseSettlementWaiters } from '../shared/promise-settlement-waiters'
+import { startWatcherGenerationReplacement } from '../main/ipc/watcher-generation-replacement'
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/relay/relay-watcher-recovery.ts]: Addressed in eb6818303. The generation replacement primitive and its test now live in src/shared, and both main runtime and relay import from that bundle-neutral boundary.

File: src/renderer/src/store/slices/hosted-review.ts [hunk 4]
```diff
@@ -369,6 +385,13 @@ export const createHostedReviewSlice: StateCreator<AppState, [], [], HostedRevie
                   repoPath,
                   ...args
                 })
+          const result = normalizeHostedReviewLookupResult(rawResult)
+          if (result.kind === 'upstream-error') {
+            hostedReviewFailureDiagnostics.report(cacheKey, repoId, result)
+            return hostedReviewFallbackAfterFailure(get, cacheKey, options?.currentHeadOid)
+          }
+          hostedReviewFailureDiagnostics.clear(cacheKey)
+          const review = result.kind === 'found' ? result.review : null
```
[the remark this answers is not in the document]
Comment (morluto) [reply src/renderer/src/store/slices/hosted-review.ts:434]: Addressed in eb6818303. Diagnostic report/clear operations are generation-gated, generic rejections no longer clear signatures, and request ownership is monotonic and HMR-persistent. Race, cross-HMR-owner, and legacy-registry migration tests cover the behavior.

**Comment by morluto** [comment]
Addressed all four CodeRabbit findings in eb6818303, including the follow-up HMR ownership edge case found during independent review. Final verification: 75 focused tests, full typecheck, staged-file lint/format hooks, and two clean independent reviews. @coderabbitai review

**Comment by brennanb2025** [comment]
Taking a look!
